bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -137,17 +147,47 @@ impl<T: AsyncRead + AsyncWrite + Unpin + Send> SourceProtocol<T> {
         self.read_ok().await
     }
 
-    async fn ram_push(&mut self) -> Result<(), MigrateError> {
-        self.update_state(MigrationState::RamPush).await;
+    async fn ram_push(
+        &mut self,
+        phase: &MigratePhase,
+    ) -> Result<(), MigrateError> {
+        // Only publish the "RAM push" phase before pausing (so that the
+        // migration doesn't go from the "pause" state back to the "RAM push"
+        // state).
+        if matches!(phase, MigratePhase::RamPushPrePause) {
+            self.update_state(MigrationState::RamPush).await;
+        }
+
         let vmm_ram_range = self.vmm_ram_bounds().await?;
         let req_ram_range = self.read_mem_query().await?;
         info!(
             self.log(),
-            "ram_push: got query for range {:?}, vm range {:?}",
+            "ram_push ({:?}): got query for range {:?}, vm range {:?}",
+            phase,
             req_ram_range,
             vmm_ram_range
         );
-        self.offer_ram(vmm_ram_range, req_ram_range).await?;
+
+        // TODO(#387): Ideally, both the pre-pause and post-pause phases would
+        // offer just dirty pages. To do this safely, the source must remember
+        // all the pages it has ever offered to any target so that they can be
+        // re-offered if migration fails and is later retried.
```

Comment:
We should probably just mark all pages as dirty in the case of a failed migration?

Reply:
Filed [illumos#15639](https://www.illumos.org/issues/15639) for this

Reply:
I think the ideal would be to have a "read but don't clear" discipline for getting dirty bits (as 15639 describes). That way we still don't have to transfer pages that have absolutely never been touched by anyone just because a prior migration failed.

That said, this seems like a useful middle ground if it turns out that migrating all pages before pausing is intractably expensive: we can just have the VM controller store the offer discipline to use, change it from "offer dirty" to "offer all" on migration failure, and then pump it into the migration context as a parameter when a new migration begins.